Add a new "txn" keyword & start using it with no functionality

mohit10verma opened the pull request:
This change adds a new Go keyword called "txn", which can be used as:
```
txn() {
 // any code
}
```
The parser/AST creator will identify this keyword & inject some statements to initialize transaction variables & begin a transaction. 
There is no other functionality yet. All the code within txn() {} will execute as they should.

To build code using txn keyword: `go build -txn`

If you use only `go build` & have txn keyword in your code, 
you get a syntax error _"syntax error: txn() use not supported. Add -txn flag while building"_

Testing: all.bash with & without -txn flag

And a simple code like this works as intended:
```
import (
	"github.com/vmware/go-pmem-transaction/pmem"
	"github.com/vmware/go-pmem-transaction/transaction"
	"fmt"
	"os"
)

func main() {
	os.Remove("testFile")
	pmem.Init("testFile")
	var a int
	txn() {
		a = 2
		fmt.Println(a)
	}

	txn() {
		b := 3
		fmt.Println(b)
	}
}
```

mohit10verma committed:
Add a new keyword called "txn"

This commit simply adds this as a known keyword. It cannot be used
yet.

mohit10verma committed:
Change parser/ast creation to identify txn keyword

With this commit a use like this will compile:
txn() {
      a = 2
      // any other type of code
}

This commit also changes parser to inject two statements to
application code:
1. to declare a transaction variable
2. Begin the transaction
This commit adds no other functionality. All other statements within
txn() block are left unchanged - so no logging, no ending of the
transaction yet.

mohit10verma committed:
Throw syntax error if txn() keyword used without txn build flag

To be able to use txn() {} in code, build with
go build -txn
Otherwise there will be a syntax error during compilation.

mohit10verma commented:
Please review each commit separately. They are related and build on top of each other.

jerrinsg commented on src/cmd/compile/internal/syntax/token_string.go:
```diff
@@ -4,9 +4,9 @@ package syntax
 import "strconv"
 
-const _token_name = "EOFnameliteralopop=opop=:=<-*([{)]},;:....breakcasechanconstcontinuedefaultdeferelsefallthroughforfuncgogotoifimportinterfacemappackagerangereturnselectstructswitchtypevar"
+const _token_name = "EOFnameliteralopop=opop=:=<-*([{)]},;:....breakcasechanconstcontinuedefaultdeferelsefallthroughforfuncgogototxnifimportinterfacemappackagerangereturnselectstructswitchtypevar"
```
The tokens seems to be lexicographically sorted. So I feel it'll be good to maintain the same order.

pratapsu replied:
I agree.

mohit10verma replied:
This file is actually auto-generated. So this can't be changed. At the top of the file, you can see the comment
`// Code generated by "stringer -type token -linecomment"; DO NOT EDIT.`

jerrinsg replied:
Is `tokens_string.go` autogenerated from `tokens.go`? If so will maintaining the order in tokens.go - 
https://github.com/jerrinsg/go-pmem/pull/10/commits/161ea731c48abdbd56c63c40aa70bbcb28a97c47#diff-dcf8ce78525d3af702e737687bc437caR56 help autogenerate this file correctly?

mohit10verma replied:
Yes. I see the order in tokens_string.go comes from the order in tokens.go. I'll fix this in the next commit. 
Thanks!

jerrinsg commented on src/cmd/compile/internal/syntax/parser.go:
```diff
@@ -2018,6 +2018,48 @@ func (p *parser) commClause() *CommClause {
+
+	var s string
+	if declTx {
+		s = "var tx transaction.TX"
```
Can we use a more convoluted variable name here (e.g. ___tx)? This will help ensure that we do not use a variable name that the user may already have in his code.

jerrinsg reviewed (commented):
The changes look good. Adding some suggestions.

pratapsu commented on src/cmd/compile/internal/gc/noder.go:
```diff
@@ -1061,6 +1071,16 @@ func (p *noder) blockStmt(stmt *syntax.BlockStmt) []*Node {
 
+func (p *noder) txBlockStmt(txB *syntax.TxBlockStmt) []*Node {
+	var nodes []*Node
+	nodes = append(nodes, p.stmts(txB.Pre)...)
```
You can merge lines 1075 and 1076.

pratapsu commented on src/cmd/compile/internal/syntax/parser.go:
```diff
@@ -2018,6 +2018,48 @@ func (p *parser) commClause() *CommClause {
+	var pTmp parser
+	src := "package p; func _() { " + s + "; }"
+	// The filename for the inserted statement is specified as tx.go
+	f, err := Parse(NewFileBase("tx.go"), strings.NewReader(src), nil, nil, 0)
```
You can merge lines 2027 and 2028.

mohit10verma replied:
Merging 2027 & 2028 will be a problem, as `f` will be out of scope when the function returns currently at line 2031.

pratapsu replied:
f will not go out of scope if you use it within the else clause. Example: 

func foo() (int, int) {
	return 4, 5
}

func main() {
	if a, b := foo(); a != 5 {
		fmt.Println(a, b)
	} else {
		fmt.Println(a+3, b)
	}
}

mohit10verma replied:
Actually in this particular instance, the next statement is a return which should be there unconditionally. I can change this to look like:
```
var s Stmt
if f, err := Parse(NewFileBase("tx.go"), strings.NewReader(src), nil, nil, 0); err != nil {
	pTmp.syntaxError("parse error in injecting statement")
} else {
	s = f.DeclList[0].(*FuncDecl).Body.List[0]
}
return s
```

Although I don't see the point. Also, I see code like this in existing compiler code. For example in cmd/compiler/internal/syntax/syntax.go:77
```
func ParseFile(filename string, errh ErrorHandler, pragh PragmaHandler, mode Mode) (*File, error) {
	f, err := os.Open(filename)
	if err != nil {
		if errh != nil {
			errh(err)
		}
		return nil, err
	}
	defer f.Close()
	return Parse(NewFileBase(filename), f, errh, pragh, mode)
}

```

pratapsu commented on src/cmd/compile/internal/syntax/parser.go:
```diff
@@ -2126,7 +2171,20 @@ func (p *parser) stmtList() (l []Stmt) {
+	firstTxKey := true
 	for p.tok != _EOF && p.tok != _Rbrace && p.tok != _Case && p.tok != _Default {
+		if p.tok == _Txn {
+			if p.mode&GenTxn == 0 {
```
If you write this as if p.mode & GenTxn, with a space before and after the &, it will be more readable.

mohit10verma replied:
This is actually auto-formatted this way by a Go tool that comes as a default with VS Code.

pratapsu commented:
For testing this change .. it is time to add a test to the test suite that uses the txn {} keyword. Then compile this with and without txn and confirm expected results.

mohit10verma committed:
Change order of keywords

So that they remain alphabetically sorted in tokens.go &
token_string.go

mohit10verma committed:
Add tests for txn keyword

The tests make sure that correct build flag must be set if txn
keyword is used, that the statements injected by parser are
correct, and that existing statements are not affected.
This commit also has other cosmetic changes related to code cleanup,
comment correction.

mohit10verma commented:
> For testing this change .. it is time to add a test to the test suite that uses the txn {} keyword. Then compile this with and without txn and confirm expected results.

I added tests for the functionality checked in with this commit. I'll add more tests as I add more functionality in the future commits.

pratapsu commented:
> > For testing this change .. it is time to add a test to the test suite that uses the txn {} keyword. Then compile this with and without txn and confirm expected results.
> 
> I added tests for the functionality checked in with this commit. I'll add more tests as I add more functionality in the future commits.

Can you share the test that you wrote? I'd like to review the tests in addition to the changes to the compiler code.

mohit10verma commented:
> > > For testing this change .. it is time to add a test to the test suite that uses the txn {} keyword. Then compile this with and without txn and confirm expected results.
> > 
> > 
> > I added tests for the functionality checked in with this commit. I'll add more tests as I add more functionality in the future commits.
> 
> Can you share the test that you wrote? I'd like to review the tests in addition to the changes to the compiler code.

Hi Pratap,

I pushed a commit today to this pull request that adds tests for these compiler changes. They are triggered by the all.bash script. This is the [commit](https://github.com/jerrinsg/go-pmem/pull/10/commits/aa704a0726186e7f8f6a5020b1ae61128907052d).

jerrinsg commented on src/cmd/compile/internal/syntax/scanner_test.go:
```diff
@@ -260,6 +260,7 @@ var sampleTokens = [...]struct {
 	{_Struct, "struct", 0, 0},
 	{_Switch, "switch", 0, 0},
 	{_Type, "type", 0, 0},
+	{_Txn, "txn", 0, 0},
```
Nit: `_Txn` should be added before `_Type` here to maintain the lexicographic order!

mohit10verma replied:
Thanks. Fixed!

pratapsu reviewed (commented):
Mohit: the test case looks ok, but where is the update to the test scripts to invoke this test with and without the tx flag?

mohit10verma commented:
Hi Pratap,

The test scripts take all the tests in a directory and run them. For the test script to actually use the txn flag the environment variable GO_GCFLAGS need to be set. I was just doing this from command line. I pushed a new [commit](https://github.com/jerrinsg/go-pmem/pull/12/commits/fbdb85047c9909c9dbee3b179e6e2c497c7a3dab) to do this from all.bash in a new pull request [here](https://github.com/jerrinsg/go-pmem/pull/12). 

Although, doing this doubles the time taken by all.bash since we are running all the tests twice.

mohit10verma committed:
Change order of txn and type keyword

jerrinsg reviewed (approved):
The changes look good to me. Thanks!
